File: LLM/bootstrap/bootstrap_rs/src/exec/install_pkg.rs
//! `install_pkg` — pip-install one package into the active venv.
//!
//! Port of `bootstrap_go/exec/install_pkg.go`. Builds the pip CLI
//! from (name, version, index, extras) and shells out via the
//! injected `Runner`.

use anyhow::{anyhow, Context, Result};
use std::path::PathBuf;
use std::time::Duration;

use crate::args::{arg_required, arg_string, arg_string_slice};
use crate::pipspec::build_pip_spec;
use crate::plan::Step;
use crate::python::python_exe_path;
use crate::runner::Runner;

pub struct InstallPkgOpts<'a> {
    pub venv_dir: PathBuf,
    pub runner: &'a dyn Runner,
}

pub fn install_pkg(opts: &InstallPkgOpts<'_>, step: &Step) -> Result<()> {
    if opts.venv_dir.as_os_str().is_empty() {
        return Err(anyhow!(
            "install_pkg: no venv configured (run create_venv first)"
        ));
    }
    let name = arg_required(&step.args, "name").context("install_pkg")?;
    let version = arg_string(&step.args, "version").unwrap_or_default();
    let index = arg_string(&step.args, "index").unwrap_or_default();
    let extras = arg_string_slice(&step.args, "extras").unwrap_or_default();

    let spec = build_pip_spec(&name, &version, &extras);
    let py = python_exe_path(&opts.venv_dir);
    let py_str = py.to_string_lossy();

    eprintln!("  install_pkg: {spec} -> {py_str}");

    // Build the arg vector. Using a Vec because the optional
    // --index-url pair makes a fixed-size array awkward.
    let mut argv: Vec<&str> = vec![
        "-m",
        "pip",
        "install",
        "--no-input",
        "--disable-pip-version-check",
        &spec,
    ];
    if !index.is_empty() {
        argv.push("--index-url");
        argv.push(&index);
    }

    opts.runner
        .run(Duration::from_secs(600), &py_str, &argv)
        .with_context(|| format!("install_pkg: pip install {spec} failed"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::runner::FakeRunner;
    use serde_json::json;

    #[test]
    fn builds_expected_pip_command() {
        let runner = FakeRunner::new();
        let step = Step::new("install_pkg").with_args(json!({
            "name": "torch",
            "version": "2.5.1+cu121",
            "index": "https://download.pytorch.org/whl/cu121",
        }));
        install_pkg(
            &InstallPkgOpts {
                venv_dir: PathBuf::from("/tmp/venv"),
                runner: &runner,
            },
            &step,
        )
        .expect("clean fake should succeed");
        let calls = runner.captured();
        assert_eq!(calls.len(), 1);
        assert_eq!(
            calls[0].args,
            vec![
                "-m",
                "pip",
                "install",
                "--no-input",
                "--disable-pip-version-check",
                "torch==2.5.1+cu121",
                "--index-url",
                "https://download.pytorch.org/whl/cu121",
            ]
        );
    }

    #[test]
    fn omits_index_when_not_given() {
        let runner = FakeRunner::new();
        let step = Step::new("install_pkg").with_args(json!({
            "name": "requests",
            "version": "2.31.0",
        }));
        install_pkg(
            &InstallPkgOpts {
                venv_dir: PathBuf::from("/tmp/venv"),
                runner: &runner,
            },
            &step,
        )
        .unwrap();
        let calls = runner.captured();
        assert!(!calls[0].args.iter().any(|a| a == "--index-url"));
        assert_eq!(calls[0].args.last().unwrap(), "requests==2.31.0");
    }

    #[test]
    fn forwards_extras_through_pip_spec() {
        let runner = FakeRunner::new();
        let step = Step::new("install_pkg").with_args(json!({
            "name": "foo",
            "version": "1.0",
            "extras": ["dev", "test"],
        }));
        install_pkg(
            &InstallPkgOpts {
                venv_dir: PathBuf::from("/tmp/venv"),
                runner: &runner,
            },
            &step,
        )
        .unwrap();
        assert_eq!(runner.captured()[0].args.last().unwrap(), "foo[dev,test]==1.0");
    }

    #[test]
    fn no_venv_fails() {
        let runner = FakeRunner::new();
        let err = install_pkg(
            &InstallPkgOpts {
                venv_dir: PathBuf::new(),
                runner: &runner,
            },
            &Step::new("install_pkg").with_args(json!({ "name": "torch" })),
        )
        .unwrap_err();
        assert!(err.to_string().contains("no venv"));
    }

    #[test]
    fn missing_name_fails() {
        let runner = FakeRunner::new();
        let err = install_pkg(
            &InstallPkgOpts {
                venv_dir: PathBuf::from("/tmp/v"),
                runner: &runner,
            },
            &Step::new("install_pkg"),
        )
        .unwrap_err();
        let chain = format!("{err:#}");
        assert!(chain.contains("missing required arg"), "got: {chain}");
    }

    #[test]
    fn runner_failure_propagates() {
        let runner = FakeRunner::failing("pip exploded");
        let err = install_pkg(
            &InstallPkgOpts {
                venv_dir: PathBuf::from("/tmp/v"),
                runner: &runner,
            },
            &Step::new("install_pkg").with_args(json!({ "name": "torch" })),
        )
        .unwrap_err();
        let chain = format!("{err:#}");
        assert!(chain.contains("pip exploded"), "got: {chain}");
    }
}

File: python-app/bootstrap/bootstrap/bootstrap_go/exec/install_pkg.go
// install_pkg: pip-install one package into the venv.
//
// Args (model-provided):
//   name:    package name (required, e.g. "torch")
//   version: optional version pin (e.g. "2.5.1+cu121")
//   index:   optional --index-url (e.g. "https://download.pytorch.org/whl/cu121")
//   extras:  optional list of extras (e.g. ["dev", "test"])
//
// Idempotence:
//   - If the package is already installed at the requested version, no-op.
//   - If installed at a different version and `version` was specified,
//     pip's resolver will upgrade/downgrade as part of the install.
//
// We intentionally use --no-input to avoid pip prompting for credentials
// or anything else (the bootstrap runs unattended).
package exec

import (
	"context"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/ruigro/LLM-Studio/LLM/bootstrap/bootstrap_go/plan"
)

// InstallPkgOpts ties install_pkg to a specific venv. The dispatcher
// owns "what venv are we using right now" -- it threads VenvDir down
// to each executor that needs a Python interpreter.
type InstallPkgOpts struct {
	VenvDir string    // required
	Runner  CmdRunner // optional; nil -> real os/exec
}

// InstallPkg builds a `python -m pip install` invocation from the args
// and runs it in the venv. Returns nil on success.
func InstallPkg(opts InstallPkgOpts, s plan.Step) error {
	if opts.VenvDir == "" {
		return fmt.Errorf("install_pkg: no venv configured (run create_venv first)")
	}
	name, err := argRequired(s.Args, "name")
	if err != nil {
		return fmt.Errorf("install_pkg: %w", err)
	}
	version, _ := argString(s.Args, "version")
	index, _ := argString(s.Args, "index")
	extras, _ := argStringSlice(s.Args, "extras")

	spec := buildPipSpec(name, version, extras)
	args := []string{"-m", "pip", "install", "--no-input", "--disable-pip-version-check", spec}
	if index != "" {
		args = append(args, "--index-url", index)
	}

	pyExe := PythonExePath(opts.VenvDir)
	log.Printf("  install_pkg: %s -> %s", spec, pyExe)

	runner := opts.Runner
	if runner == nil {
		runner = realRunner
	}
	if err := runner(context.Background(), 600*time.Second, pyExe, args...); err != nil {
		return fmt.Errorf("install_pkg: pip install %s failed: %w", spec, err)
	}
	return nil
}

// buildPipSpec composes a pip requirement string. Public for unit tests.
//
// Examples:
//   ("torch", "",         nil)        -> "torch"
//   ("torch", "2.5.1",    nil)        -> "torch==2.5.1"
//   ("torch", "2.5.1+cu121", nil)     -> "torch==2.5.1+cu121"
//   ("foo",   "",         []{"a","b"})-> "foo[a,b]"
//   ("foo",   "1.0",      []{"a"})    -> "foo[a]==1.0"
//
// The model occasionally encodes the version in the name itself
// ("torch==2.5.1+cu121"). We detect that and don't double-suffix.
func buildPipSpec(name, version string, extras []string) string {
	hasInlineVersion := strings.ContainsAny(name, "=<>~!")
	if hasInlineVersion {
		return name
	}
	out := name
	if len(extras) > 0 {
		out += "[" + strings.Join(extras, ",") + "]"
	}
	if version != "" {
		out += "==" + version
	}
	return out
}
